From 1a31957af68baeba1065a250382e4744709b80a6 Mon Sep 17 00:00:00 2001
From: Ben Cressey <bcressey@amazon.com>
Date: Thu, 14 Nov 2024 19:55:38 +0000
Subject: [PATCH] tsc: drop tsc_boot_time offset

To implement the Boot Loader Interface, we need to store the time at
which GRUB was started in the `LoaderTimeInitUSec` variable.

On the i386 architecture, GRUB stores the TSC value at the time that
TSC calibration took place, and subtracts it before returning a value
from grub_tsc_get_time_ms(), which is called by grub_get_time_ms() -
the architecture-independent function for retrieving timestamps.

Storing the TSC value at the time of initialization, and subtracting
it from subsequent calculations, prevents callers from determining
the system time elapsed before GRUB was loaded.

The equivalent offset is not done for arm64, and does not appear to
serve a purpose. Callers of grub_get_time_ms() all follow a pattern
where they first record a "start" time and then compute a delta in a
loop, which works the same way whether the "start" time is close to
zero or not.

Rather than exposing another function to provide the "raw" timestamp
without the offset, just drop `tsc_boot_time` from the calculation.

Signed-off-by: Ben Cressey <bcressey@amazon.com>
---
 grub-core/kern/i386/tsc.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/grub-core/kern/i386/tsc.c b/grub-core/kern/i386/tsc.c
index 9293b161d..ec6899812 100644
--- a/grub-core/kern/i386/tsc.c
+++ b/grub-core/kern/i386/tsc.c
@@ -25,9 +25,6 @@
 #include <grub/i386/tsc.h>
 #include <grub/i386/cpuid.h>
 
-/* This defines the value TSC had at the epoch (that is, when we calibrated it). */
-static grub_uint64_t tsc_boot_time;
-
 /* Calibrated TSC rate.  (In ms per 2^32 ticks) */
 /* We assume that the tick is less than 1 ms and hence this value fits
    in 32-bit.  */
@@ -36,7 +33,7 @@ grub_uint32_t grub_tsc_rate;
 static grub_uint64_t
 grub_tsc_get_time_ms (void)
 {
-  grub_uint64_t a = grub_get_tsc () - tsc_boot_time;
+  grub_uint64_t a = grub_get_tsc ();
   grub_uint64_t ah = a >> 32;
   grub_uint64_t al = a & 0xffffffff;
 
@@ -63,8 +60,6 @@ grub_tsc_init (void)
       return;
     }
 
-  tsc_boot_time = grub_get_tsc ();
-
 #if defined (GRUB_MACHINE_XEN) || defined (GRUB_MACHINE_XEN_PVH)
   (void) (grub_tsc_calibrate_from_xen () || calibrate_tsc_hardcode());
 #elif defined (GRUB_MACHINE_EFI)
-- 
2.47.0

